File: code-library/09-data-cleaning/data-cleaning.py
# ==============================================================================
#  MODULE 9: DATA CLEANING & VALIDATION
#  Dataset: CES 2020 (cleaned in Module 1)
#
#  Stack: pandas, numpy, missingno (missing data viz)
#  pip install missingno
# ==============================================================================

import pandas as pd
import numpy as np
import missingno as msno
import matplotlib.pyplot as plt
from scipy import stats

ces = pd.read_parquet("CES2020_clean.parquet")


# ==============================================================================
# SECTION 1: INITIAL AUDIT
# ==============================================================================

print("Shape:", ces.shape)
print("\nColumn types:\n", ces.dtypes.head(20))
print("\nFirst 3 rows:\n", ces.head(3))

# pandas describe() gives the five-number summary for numerics
# For a full skim-like summary, use the pandas-profiling/ydata-profiling package:
#   pip install ydata-profiling
#   from ydata_profiling import ProfileReport
#   profile = ProfileReport(ces); profile.to_file("ces_profile.html")

print("\nDescriptive stats (selected vars):")
print(ces[["age", "education", "ideology5", "imm_restrict", "econ_retro"]].describe().round(2))


# ==============================================================================
# SECTION 2: MISSING DATA AUDIT
# ==============================================================================

# Proportion missing for every column
missing_summary = (
    ces.isna()
    .mean()
    .mul(100)
    .round(2)
    .rename("pct_missing")
    .reset_index()
    .rename(columns={"index": "variable"})
    .sort_values("pct_missing", ascending=False)
)

print("\nMissing data (all columns with any NA):")
print(missing_summary[missing_summary["pct_missing"] > 0].to_string(index=False))

# Visualize missing data patterns
key_vars = ["age", "education", "ideology5", "party_id3",
            "imm_restrict", "voted", "biden_voter", "white_nh", "college"]

fig, ax = plt.subplots(figsize=(10, 5))
msno.matrix(ces[key_vars], ax=ax, sparkline=False)
ax.set_title("Missing Data Pattern — Key Analysis Variables")
plt.tight_layout()
plt.savefig("missing_data_pattern.png", dpi=300)
plt.close()

# Bar chart of missing proportions
msno.bar(ces[key_vars], figsize=(10, 5), fontsize=10, color="navy")
plt.title("Proportion Non-Missing by Variable")
plt.tight_layout()
plt.savefig("missing_data_bar.png", dpi=300)
plt.close()

# Check whether missingness on ideology5 clusters in certain groups
print("\n% missing ideology5 by party_id3:")
print(
    ces.assign(miss_ideo=ces["ideology5"].isna())
    .groupby("party_id3")["miss_ideo"]
    .mean()
    .mul(100)
    .round(2)
)


# ==============================================================================
# SECTION 3: LOGICAL CONSISTENCY CHECKS
# ==============================================================================

# Collect all checks in a DataFrame for a clean summary report

checks = {}

# Age plausibility
checks["Age < 18 or > 110"] = (
    (ces["age"] < 18) | (ces["age"] > 110)
).sum()

# Biden voter but didn't vote
checks["Biden voter but voted == 0"] = (
    ces["biden_voter"].notna() & (ces["voted"] == 0)
).sum()

# College recode consistency
checks["college=1 but educ < 5"] = (
    (ces["college"] == 1) & (ces["education"] < 5)
).sum()
checks["college=0 but educ >= 5"] = (
    (ces["college"] == 0) & (ces["education"] >= 5)
).sum()

# White non-Hispanic consistency
checks["white_nh error"] = (
    (ces["white_nh"] == 1) & ~((ces["race_eth"] == 1) & (ces["hispanic_id"] == 2))
).sum()

# Party ID consistency
checks["Strong Rep coded as Dem"] = (
    (ces["party_id7"] == 7) & (ces["party_id3"] == 1)
).sum()

check_df = pd.DataFrame.from_dict(checks, orient="index", columns=["n_flagged"])
check_df["pass"] = check_df["n_flagged"] == 0

print("\nLogical Consistency Checks:")
print(check_df.to_string())

# Assert all checks pass (uncomment in production)
# assert check_df["pass"].all(), "Consistency checks failed"


# ==============================================================================
# SECTION 4: DUPLICATE ID CHECK
# ==============================================================================

# duplicated() returns True for every row that is a duplicate
dupes = ces[ces.duplicated(subset=["caseid"], keep=False)]
print(f"\nDuplicate caseids: {len(dupes):,} (should be 0)")

# Value counts — any caseid appearing more than once?
id_counts = ces["caseid"].value_counts()
print(f"Max times any caseid appears: {id_counts.max()}")

assert len(dupes) == 0, "Duplicate caseids found"


# ==============================================================================
# SECTION 5: OUTLIER DETECTION
# ==============================================================================

# Z-score method: |z| > 4 for large N
z_age = (ces["age"] - ces["age"].mean()) / ces["age"].std()
high_z = ces[(z_age.abs() > 4) & z_age.notna()]
print(f"\nAge outliers (|z| > 4): {len(high_z):,}")
print(high_z[["caseid", "age"]].head(10))

# IQR method (robust to extreme values)
q1, q3 = ces["age"].quantile([0.25, 0.75])
iqr     = q3 - q1
lo, hi  = q1 - 3 * iqr, q3 + 3 * iqr
iqr_out = ces[(ces["age"] < lo) | (ces["age"] > hi)]
print(f"Age IQR outliers (3×IQR): {len(iqr_out):,}")


# ==============================================================================
# SECTION 6: HANDLING MISSING DATA
# ==============================================================================

# LISTWISE DELETION (default in statsmodels)
import statsmodels.formula.api as smf

m = smf.ols("imm_restrict ~ education + age + ideology5", data=ces).fit()
print(f"\nModel N: {int(m.nobs):,} | Total N: {len(ces):,} | "
      f"Dropped: {len(ces) - int(m.nobs):,}")

# MEAN IMPUTATION (mechanics demo — not recommended for inference)
ces_imp = ces.copy()
ces_imp["ideology5_imp"] = ces["ideology5"].fillna(ces["ideology5"].mean())

print(f"\nMean imputed ideology5: "
      f"mean={ces_imp['ideology5_imp'].mean():.3f}, "
      f"sd={ces_imp['ideology5_imp'].std():.3f} "
      f"(vs original: sd={ces['ideology5'].std():.3f})")
# SD decreases after mean imputation — a known problem

del ces_imp

# FORWARD FILL (for time-ordered panel data — not appropriate here,
# shown for reference)
# ces_panel_sorted.sort_values(["caseid","wave"]).groupby("caseid")["ideology5"].ffill()

# MULTIPLE IMPUTATION — see Module 10 (using sklearn IterativeImputer)


# ==============================================================================
# SECTION 7: RECODING VARIABLES
# ==============================================================================

# Age groups with pd.cut()
ces["age_group"] = pd.cut(
    ces["age"],
    bins   = [17, 29, 44, 59, 200],
    labels = ["18-29", "30-44", "45-59", "60+"],
    right  = True
)
print("\nAge groups:\n", ces["age_group"].value_counts().sort_index())

# Ideology collapsed to 3 categories using pd.cut
ces["ideology3"] = pd.cut(
    ces["ideology5"],
    bins   = [0, 2, 3, 5],
    labels = ["Liberal", "Moderate", "Conservative"],
    right  = True
)
print("\nIdeology3:\n", ces["ideology3"].value_counts())

# Quartiles with pd.qcut()
ces["educ_quartile"] = pd.qcut(
    ces["education"],
    q      = 4,
    labels = ["Q1", "Q2", "Q3", "Q4"],
    duplicates = "drop"
)
print("\nEducation quartiles:\n", ces["educ_quartile"].value_counts().sort_index())


# ==============================================================================
# SECTION 8: STANDARDIZING VARIABLES (Z-SCORES)
# ==============================================================================

# scipy.stats.zscore() or manual normalization
for col in ["age", "education", "ideology5", "imm_restrict", "econ_retro"]:
    ces[f"z_{col}"] = (ces[col] - ces[col].mean()) / ces[col].std()

print("\nStandardized variable means (should be ~0):")
z_cols = [c for c in ces.columns if c.startswith("z_")]
print(ces[z_cols].mean().round(6))

print("Standardized variable SDs (should be ~1):")
print(ces[z_cols].std().round(6))

# Regression with standardized predictors
m_std = smf.ols(
    "z_imm_restrict ~ z_education + z_age + z_ideology5 + C(sex)",
    data=ces
).fit()
print("\nStandardized coefficients:\n", m_std.params.round(4))

# Drop z-score columns
ces = ces.drop(columns=z_cols)


# ==============================================================================
# SECTION 9: STRING CLEANING FOR MERGE KEYS
# ==============================================================================

# Simulate a state name column (as you'd get from a messy CSV)
state_map = {12:"Florida", 13:"Georgia", 36:"New York",
             48:"texas", 6:" California ", 42:"Pennsylvania"}

ces["state_name_raw"] = ces["state_fips"].map(state_map)

# Standardize: strip, lower, replace spaces
ces["state_key"] = (
    ces["state_name_raw"]
    .str.strip()
    .str.lower()
    .str.replace(r"\s+", " ", regex=True)
)

print("\nSample state keys:")
print(ces[["state_fips", "state_name_raw", "state_key"]].dropna().head(10))

ces = ces.drop(columns=["state_name_raw", "state_key"])


# ==============================================================================
# SECTION 10: ASSERTING DATA QUALITY BEFORE ANALYSIS
# ==============================================================================

# Production-grade validation: assert statements stop execution if violated.
# Run these at the start of any analysis script that depends on cleaned data.

def validate_ces(df):
    """Assert data quality constraints for CES analysis dataset."""
    # ID is unique
    assert df["caseid"].nunique() == len(df), "Duplicate caseids"

    # Age is plausible
    assert (df["age"].dropna() >= 18).all(), "Respondent younger than 18"
    assert (df["age"].dropna() <= 115).all(), "Respondent older than 115"

    # Binary vars are only 0/1/NaN
    for col in ["voted", "college", "white_nh", "dem", "rep", "biden_voter"]:
        vals = df[col].dropna().unique()
        assert set(vals).issubset({0, 1}), f"{col} has values outside {{0,1}}: {vals}"

    # imm_restrict is bounded [0, 5]
    valid = df["imm_restrict"].dropna()
    assert (valid >= 0).all() and (valid <= 5).all(), "imm_restrict out of [0,5] range"

    # biden_voter is only among voters
    assert (df.loc[df["biden_voter"].notna(), "voted"] == 1).all(), \
        "biden_voter coded for non-voters"

    print("All validation checks passed.")

validate_ces(ces)


# ==============================================================================
# SECTION 11: SAVE ANALYSIS-READY DATASET
# ==============================================================================

ces.to_parquet("CES2020_analysis_ready.parquet", index=False)
ces.to_csv("CES2020_analysis_ready.csv", index=False)

print("\nModule 9 complete. Analysis-ready dataset saved.")

File: code-library/09-data-cleaning/data-cleaning.R
# ==============================================================================
#  MODULE 9: DATA CLEANING & VALIDATION
#  Dataset: CES 2020 (raw, before Module 1 cleaning)
#
#  Purpose: Systematic data cleaning workflow — detecting and handling
#  missing data, outliers, inconsistent responses, duplicates, and
#  string-to-numeric conversions using dplyr and tidyr.
# ==============================================================================

library(tidyverse)
library(janitor)     # tabyl(), clean_names(), get_dupes() — install.packages("janitor")
library(visdat)      # visualize missing data patterns — install.packages("visdat")
library(naniar)      # tidy missing data tools — install.packages("naniar")

ces <- readRDS("CES2020_clean.rds")


# ==============================================================================
# SECTION 1: INITIAL AUDIT
# ==============================================================================

glimpse(ces)            # variable names, types, first few values
dim(ces)                # rows x columns
skimr::skim(ces)        # comprehensive univariate summary (install.packages("skimr"))


# ==============================================================================
# SECTION 2: MISSING DATA AUDIT
# ==============================================================================

# Count and proportion missing for every variable
missing_summary <- ces |>
  summarise(across(everything(),
                   list(n_miss = ~ sum(is.na(.x)),
                        pct    = ~ mean(is.na(.x)) * 100))) |>
  pivot_longer(everything(),
               names_to  = c("variable", ".value"),
               names_sep = "__") |>
  arrange(desc(pct)) |>
  filter(pct > 0)

print(missing_summary, n = 30)

# Which key analysis variables have the most missing?
ces |>
  select(age, education, ideology5, party_id3, imm_restrict,
         voted, biden_voter, white_nh, college) |>
  summarise(across(everything(), ~ mean(is.na(.x)) * 100)) |>
  pivot_longer(everything(), names_to = "var", values_to = "pct_missing") |>
  arrange(desc(pct_missing))

# Visualize missing data patterns
vis_miss(
  ces |> select(age, education, ideology5, party_id3, imm_restrict,
                voted, biden_voter, white_nh, college),
  sort_miss = TRUE
)

# Upset plot: which variables tend to be missing together?
gg_miss_upset(
  ces |> select(age, education, ideology5, party_id3, imm_restrict,
                voted, biden_voter)
)


# ==============================================================================
# SECTION 3: LOGICAL CONSISTENCY CHECKS
# ==============================================================================

# Impossible ages
ces |> filter(age < 18 | age > 110) |> nrow()
ces |> filter(age < 18 | age > 110) |> select(caseid, birth_year, age)

# biden_voter should be missing unless voted == 1
# Anyone with biden_voter coded but voted == 0?
ces |>
  filter(!is.na(biden_voter), voted == 0) |>
  nrow()

# College recode check: college==1 requires education >= 5
ces |>
  filter(college == 1 & education < 5) |>
  nrow()   # should be 0

ces |>
  filter(college == 0 & education >= 5) |>
  nrow()   # should be 0

# White non-Hispanic check: white_nh==1 requires race_eth==1 AND hispanic_id==2
ces |>
  filter(white_nh == 1 & (race_eth != 1 | hispanic_id != 2)) |>
  nrow()   # should be 0

# Party ID consistency: Strong Republican on pid7 but Democrat on pid3?
ces |>
  filter(party_id7 == 7 & party_id3 == 1) |>
  nrow()

ces |>
  filter(party_id7 == 1 & party_id3 == 2) |>
  nrow()

# Report all check results in one tidy block
checks <- tibble(
  check         = c("Age < 18 or > 110",
                    "Biden voter but voted == 0",
                    "College=1 but educ < 5",
                    "College=0 but educ >= 5",
                    "white_nh error",
                    "Strong Rep coded as Dem"),
  n_flagged     = c(
    sum((ces$age < 18 | ces$age > 110), na.rm = TRUE),
    sum(!is.na(ces$biden_voter) & ces$voted == 0, na.rm = TRUE),
    sum(ces$college == 1 & ces$education < 5, na.rm = TRUE),
    sum(ces$college == 0 & ces$education >= 5, na.rm = TRUE),
    sum(ces$white_nh == 1 & (ces$race_eth != 1 | ces$hispanic_id != 2), na.rm = TRUE),
    sum(ces$party_id7 == 7 & ces$party_id3 == 1, na.rm = TRUE)
  )
)

checks |> mutate(pass = n_flagged == 0)


# ==============================================================================
# SECTION 4: DUPLICATE ID CHECK
# ==============================================================================

# Check for duplicate caseids using janitor::get_dupes()
get_dupes(ces, caseid)   # returns only rows that are duplicates

# Manual approach
ces |>
  count(caseid) |>
  filter(n > 1) |>
  nrow()   # should be 0 for a cross-sectional survey


# ==============================================================================
# SECTION 5: OUTLIER DETECTION — CONTINUOUS VARIABLES
# ==============================================================================

# Z-score approach: flag |z| > 4 (appropriate for large N)
outlier_flags <- ces |>
  mutate(
    z_age = (age - mean(age, na.rm=TRUE)) / sd(age, na.rm=TRUE)
  ) |>
  filter(abs(z_age) > 4) |>
  select(caseid, age, z_age) |>
  arrange(desc(abs(z_age)))

outlier_flags |> head(20)

# IQR method (more robust to extreme values)
q <- quantile(ces$age, c(0.25, 0.75), na.rm = TRUE)
iqr <- q[2] - q[1]
lo  <- q[1] - 3 * iqr
hi  <- q[2] + 3 * iqr

ces |> filter(age < lo | age > hi) |> nrow()


# ==============================================================================
# SECTION 6: HANDLING MISSING DATA — STRATEGIES
# ==============================================================================

# STRATEGY 1: LISTWISE DELETION (default in lm/glm)
# R's lm() drops NA rows automatically — check model N vs total N
m <- lm(imm_restrict ~ education + age + ideology5, data = ces)
cat("Model N:", nobs(m), "| Total N:", nrow(ces), "\n")

# STRATEGY 2: SIMPLE MEAN IMPUTATION (demonstrating mechanics, not recommended)
ces_mean_imp <- ces |>
  mutate(
    ideology5_imp = if_else(
      is.na(ideology5),
      mean(ideology5, na.rm = TRUE),
      as.double(ideology5)
    )
  )

# Compare distributions
ces |> summarise(mean = mean(ideology5, na.rm=TRUE), sd = sd(ideology5, na.rm=TRUE))
ces_mean_imp |> summarise(mean = mean(ideology5_imp), sd = sd(ideology5_imp))
# SD is lower after mean imputation — underestimates variance

rm(ces_mean_imp)

# STRATEGY 3: MULTIPLE IMPUTATION — see Module 10


# ==============================================================================
# SECTION 7: RECODING VARIABLES
# ==============================================================================

# Replace specific values with NA (more targeted than across())
ces <- ces |>
  mutate(
    econ_retro = na_if(econ_retro, 6)   # 6 = "Not sure" → missing
  )

# Collapsing categories
ces <- ces |>
  mutate(
    # Age groups with meaningful sociological categories
    age_group = case_when(
      age < 30              ~ "18-29",
      age >= 30 & age < 45  ~ "30-44",
      age >= 45 & age < 60  ~ "45-59",
      age >= 60             ~ "60+",
      TRUE                  ~ NA_character_
    ) |> factor(levels = c("18-29", "30-44", "45-59", "60+")),

    # Ideology collapsed to 3 categories
    ideology3 = case_when(
      ideology5 %in% 1:2 ~ "Liberal",
      ideology5 == 3     ~ "Moderate",
      ideology5 %in% 4:5 ~ "Conservative",
      TRUE               ~ NA_character_
    ) |> factor(levels = c("Liberal", "Moderate", "Conservative")),

    # Education quartile
    educ_quartile = ntile(education, 4)
  )

ces |> count(age_group)
ces |> count(ideology3)
ces |> count(educ_quartile)


# ==============================================================================
# SECTION 8: STANDARDIZING VARIABLES (Z-SCORES)
# ==============================================================================

# scale() centers and divides by SD — creates a numeric matrix
# as.numeric() extracts the vector from the matrix wrapper

ces <- ces |>
  mutate(
    z_age       = as.numeric(scale(age)),
    z_education = as.numeric(scale(education)),
    z_ideology  = as.numeric(scale(ideology5)),
    z_restrict  = as.numeric(scale(imm_restrict)),
    z_econ      = as.numeric(scale(econ_retro))
  )

# Standardized regression
lm(z_restrict ~ z_education + z_age + z_ideology + factor(sex),
   data = ces) |>
  broom::tidy()

# Drop z-score columns after use (optional)
ces <- ces |> select(-starts_with("z_"))


# ==============================================================================
# SECTION 9: CLEAN VARIABLE NAMES WITH JANITOR
# ==============================================================================

# clean_names() converts all names to snake_case and handles special characters
# Useful after importing messy CSV headers

demo_messy <- tibble(
  `Respondent ID` = 1:5,
  `Age (Years)`   = c(35, 42, 28, 51, 67),
  `Party.ID`      = c(1, 2, 3, 1, 2),
  `VoteChoice2020` = c(1, 0, 1, 1, 0)
)

demo_clean <- demo_messy |> clean_names()
names(demo_clean)   # respondent_id, age_years, party_id, votechoice2020


# ==============================================================================
# SECTION 10: TABYL — JANITOR'S FREQUENCY TABLES
# ==============================================================================

# tabyl() is a tidyverse-friendly alternative to table()
# Handles NA explicitly and formats output cleanly

ces |> tabyl(sex) |> adorn_pct_formatting()
ces |> tabyl(voted) |> adorn_pct_formatting()
ces |> tabyl(college) |> adorn_pct_formatting()

# Two-way tabyl with chi-square test
ces |>
  tabyl(college, biden_voter) |>
  adorn_percentages("row") |>
  adorn_pct_formatting() |>
  adorn_ns()   # add raw counts in parentheses

ces |>
  tabyl(party_id3, biden_voter) |>
  chisq.test()


# ==============================================================================
# SECTION 11: FINAL SAVE
# ==============================================================================

# Add new variables and save the analysis-ready file
saveRDS(ces, "CES2020_analysis_ready.rds")
write_csv(ces, "CES2020_analysis_ready.csv")

message("Module 9 complete. Analysis-ready dataset saved.")
